src/core/public/context/context.ts

```diff
@@ -0,0 +1,222 @@
+   * lifecycle it will throw an exception.
+   *
+   * @param handler
+   */
+  createHandler(
+    handler: Handler<TContext, THandlerReturn, THandlerParameters>
+  ): (...rest: THandlerParameters) => Promisify<THandlerReturn>;
+}
+
+/** @internal */
+export class ContextContainerImplementation<
+  TContext extends {},
+  THandlerReturn,
+  THandlerParameters extends any[] = []
+> implements ContextContainer<TContext, THandlerReturn, THandlerParameters> {
+  private currentPlugin?: string;
+
+  /**
+   * Used to map contexts to their providers and associated plugin. In registration order which is tightly coupled to
+   * plugin load order.
+   */
+  private readonly contextProviders = new Map<
+    keyof TContext,
+    {
+      provider: ContextProvider<TContext, keyof TContext, THandlerParameters>;
+      plugin?: PluginName;
+    }
+  >();
+  /** Used to keep track of which plugins registered which contexts for dependency resolution. */
+  private readonly contextNamesByPlugin = new Map<PluginName, Array<keyof TContext>>();
+
+  /**
+   * @param pluginDependencies - A map of plugins to an array of their dependencies.
+   */
+  constructor(private readonly pluginDependencies: ReadonlyMap<PluginName, PluginName[]>) {}
+
+  public registerContext = <TContextName extends keyof TContext>(
+    contextName: TContextName,
+    provider: ContextProvider<TContext, TContextName, THandlerParameters>
+  ): this => {
+    if (this.contextProviders.has(contextName)) {
+      throw new Error(`Context provider for ${contextName} has already been registered.`);
+    }
+
+    const plugin = this.currentPlugin;
+    this.contextProviders.set(contextName, { provider, plugin });
+
+    if (plugin) {
+      this.contextNamesByPlugin.set(plugin, [
+        ...(this.contextNamesByPlugin.get(plugin) || []),
+        contextName,
+      ]);
+    }
+
+    return this;
+  };
+
+  public createHandler = (handler: Handler<TContext, THandlerReturn, THandlerParameters>) => {
+    const plugin = this.currentPlugin;
+    if (!plugin) {
+      throw new Error(`Cannot create handlers outside a plugin!`);
+    } else if (!this.pluginDependencies.has(plugin)) {
+      throw new Error(`Cannot create handler for unknown plugin: ${plugin}`);
+    }
+
+    return (async (...args: THandlerParameters) => {
+      const context = await this.buildContext(plugin, {}, ...args);
+      return handler(context, ...args);
+    }) as (...args: THandlerParameters) => Promisify<THandlerReturn>;
+  };
+
+  private async buildContext(
+    pluginName: PluginName,
+    baseContext: Partial<TContext> = {},
+    ...contextArgs: THandlerParameters
+  ): Promise<TContext> {
+    const ownerContextNames = [...this.contextProviders]
+      .filter(([name, { plugin }]) => plugin === undefined)
+      .map(([name]) => name);
+    const contextNamesForPlugin = (plug: PluginName): Set<keyof TContext> => {
```

Comment:
could you extract it in a separate method? imo it's confusing to read and thinking about scope of the function